docs(core): node and react tutorial rework (#12498)
nrwl/nx: docs/shared/react-tutorial/4-workspace-optimization.md

```bash
npx nx affected --target=test
```

This will run the `test` target for all projects that:

1. are affected by our change, and
2. have a `test` target in their `project.json` file.

This can be particularly helpful in CI pipelines for larger repos, where most commits only affect a small subset of the entire workspace.

{% callout type="note" title="Affected API Documentation" %}
See the [Nx API documentation for affected](nx/affected) for full details.
{% /callout %}

## Task Caching

Task Caching is another mechanism to optimize your workspace.

Whereas `affected` allows you to "skip" tasks that couldn't possibly be affected by your changes, Task Caching allows you to "replay" tasks that had already been run.

Also, while `affected` is informed by your git metadata, Task Caching is informed by "inputs" and "outputs":

### Inputs

Inputs for your task caching includes by default any environment details and all the source code of the projects and dependencies affecting your project.

> When running a task, Nx will determine all the inputs for your project and create a hash that will be used to index your cache. If you've already run this task with the same inputs, your cache will already be populated for at this index, and Nx will replay the results stored in the cache.

Reviewer: ```suggestion
When running a task, Nx will determine all the inputs for your project and create a hash that will be used to index your cache. If you've already run this task with the same inputs, your cache will already be populated at this index, and Nx will replay the results stored in the cache.
```

If this index does not exist, Nx will actually run the command and then store those results in the cache.

### Outputs

> Outputs of the cache include the terminal output craeted by the task, as well as any files created by the task - for example: the artifact created by running a `build` task.

Reviewer: ```suggestion
Outputs of the cache include the terminal output created by the task, as well as any files created by the task - for example: the artifact created by running a `build` task.
```

Entrypoint polyfills 93.5 KiB = runtime.54e36ebee261465d.js 1.19 KiB polyfills.bd0d0abec287a28e.js 92.3 KiB
Entrypoint styles 1.19 KiB = runtime.54e36ebee261465d.js 1.19 KiB styles.ef46db3751d8e999.css 0 bytes
chunk (runtime: runtime) main.623d91691bdb6754.css, main.303fe7c1dcf5306b.js (main) 144 KiB (javascript) 50 bytes (css/mini-extract) [initial] [rendered]
chunk (runtime: runtime) polyfills.bd0d0abec287a28e.js (polyfills) 301 KiB [initial] [rendered]
chunk (runtime: runtime) styles.ef46db3751d8e999.css (styles) 50 bytes (javascript) 80 bytes (css/mini-extract) [initial] [rendered]
chunk (runtime: runtime) runtime.54e36ebee261465d.js (runtime) 3.23 KiB [entry] [rendered]
webpack compiled successfully (0c0df3e6c70c6b7b)

 ———————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Successfully ran target build for project admin (4s)
```

> Since you had not run the `build` target before for the `admin` project, Nx ran the `build`, and populated the results in `dist/apps/admin` as specified in the `admin` project's `project.json` file for the `build` target, which we can see took 4 seconds.

Reviewer: ```suggestion
Since you have not run the `build` target before for the `admin` project, Nx runs the `build`, and populates the results in `dist/apps/admin` as specified in the `admin` project's `project.json` file for the `build` target, which we can see takes 4 seconds.
```

webpack compiled successfully (06e95dfdacea84c7)

 ———————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Successfully ran target build for project store and 1 task(s) it depends on (5s)
```

Notice the line here:

```bash
   ✔    1/1 dependent project tasks succeeded [0 read from cache]
```

This is because your `store` project is dependent on your `products` project, which is also buildable. By default Nx is configured to run (or pull results from cache) for any dependent project's `build`s.

{% callout type="note" title="Customizing Task Inputs" %}
When you generated your workspace, those scripts set reasonable `targetDefaults` in your `nx.json` file.

See our [guide for Customizing Inputs and Named Inputs](/more-concepts/customizing-inputs) for more information on how to customize these to better fit your workspace.
{% /callout %}

> This can be particularly helpful in relying on your Nx graph to dynamically maintain task dependencies, rather than having to manually maintain those task dependencies as your workspace continues to grow.

Reviewer: ```suggestion
This feature allows the Nx graph to dynamically maintain task dependencies, rather than having to manually maintain those task dependencies as your workspace continues to grow.
```
